storey/writers.py

```diff
@@ -84,12 +84,16 @@ def _event_to_writer_entry(self, event):
             self._get_column_data_from_dict(data, event, self._columns, self._metadata_columns, self._rename_columns)
             self._get_column_data_from_dict(data, event, self._index_cols, self._metadata_index_columns, self._rename_index_columns)
         elif isinstance(data, list):
+            if self._infer_columns_from_data:
+                raise TypeError('Cannot infer_columns_from_data when event type is list. Inference is only possible from dict.')
             sub_metadata = bool(self._columns) and bool(self._metadata_columns)
             sub_index_metadata = bool(self._index_cols) and bool(self._metadata_index_columns)
             if sub_metadata or sub_index_metadata:
                 data = []
                 cursor = self._get_column_data_from_list(data, event, event.body, self._index_cols, self._metadata_index_columns)
                 self._get_column_data_from_list(data, event, event.body[cursor:], self._columns, self._metadata_columns)
+        else:
+            raise TypeError('Writer supports only events of type dict or list.')
```

Comment:
better use the step's name rather than "Writer".
a flow can have multiple branches so the error should be clear enough

Reply:
Actually, we better add the step name to every error we raise.. but that's for later